enhancements/ingress/set_eip_nlb_ingress.md

```diff
@@ -0,0 +1,983 @@
+
+```go
+// Platform stores all the global configuration that all machinesets
+// use.
+type Platform struct {
+...
+
+    // networkLoadBalancerParameters holds configuration parameters for an AWS
+    // network load balancer. Present only if type is NLB.
+    //
+    // +optional
+    NetworkLoadBalancerParameters *configv1.AWSNetworkLoadBalancerParameters `json:"nlbParameters,omitempty"`
+}
+
+```
+
+- The installer will propagate `eipAllocations` from `install-config.yaml` to the cluster ingress configuration, which the ingress operator uses to create the default IngressController.  
+  The following  definitions will be added in the file [`config/v1/types_ingress.go`](https://github.com/openshift/api/blob/master/config/v1/types_ingress.go) of the [`openshift/api`](https://github.com/openshift/api) GitHub repository:
+
+```go
+// AWSIngressSpec holds the desired state of the Ingress for Amazon Web Services infrastructure provider.
+// This only includes fields that can be modified in the cluster.
+// +kubebuilder:validation:XValidation:rule="self.type != 'Classic' || !has(self.nlbParameters)",message="Network load balancer parameters are allowed only when load balancer type is NLB."
+// +union
+type AWSIngressSpec struct {
+...
+    // +unionDiscriminator
+    // +kubebuilder:validation:Enum:=NLB;Classic
+    // +kubebuilder:validation:Required
+    Type AWSLBType `json:"type,omitempty"`
+
+    // networkLoadBalancerParameters holds configuration parameters for an AWS
+    // network load balancer. Present only if type is NLB.
+    //
+    // +optional
+    NetworkLoadBalancerParameters *AWSNetworkLoadBalancerParameters `json:"nlbParameters,omitempty"`
+}
+
+// AWSNetworkLoadBalancerParameters holds configuration parameters for an
+// AWS Network load balancer. For Example: Setting AWS EIPs https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html
+type AWSNetworkLoadBalancerParameters struct {
+    // You can assign Elastic IP addresses to the Network Load Balancer by adding the following annotation.
+    // The number of Allocation IDs must match the number of subnets that are used for the load balancer.
+    // service.beta.kubernetes.io/aws-load-balancer-eip-allocations: eipalloc-xxxxxxxxxxxxxxxxx,eipalloc-yyyyyyyyyyyyyyyyy
+    // https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/elastic-ip-addresses-eip.html
+    // https://kubernetes-sigs.github.io/aws-load-balancer-controller/v2.7/guide/service/annotations/#eip-allocations
+    // +openshift:enable:FeatureGate=SetEIPForNLBIngressController 
+    // The EIPs in the cluster config are only used when the operator creates the default IngressController. 
+    // It isn't used as a default value when eipAllocations is not specified on the IngressController.
```

Comment:
I think this is a important difference that concerns me about the Ingress object's API consistency regarding `default` IngressController vs. default value for new IngressControllers.

Reviewing the current fields:

- `spec.loadBalancer.platform.aws.type`:
  - Sets the LBType for the `default` IngressController AND the default value for new IngressControllers (note: the [docs](https://github.com/openshift/api/blob/9fb7ffb45a2d501f28d850e394fc597c3a5dca29/config/v1/types_ingress.go#L137-L139) are a bit misleading, as it doesn't point out the fact that it also sets the default value.
- `spec.loadBalancer.platform.aws.nlbParameters.eipAllocations`: [this EP's API]
  - Sets the eipAllocations for `default` IngressController *only*.
- `spec.loadBalancer.platform.aws.subnets`: [my draft [EP](https://github.com/openshift/enhancements/pull/1634)]
  - Sets the subnets for the `default` IngressController AND the default value for new IngressControllers.

Now `eipAllocations` use case *only* set the `default` IngressController makes sense. Setting the default value for new IngressControllers would be a bit silly, since you can't share EIPs. But this field is inconsistent with the other fields surrounding it. That seems potentially confusing. 

My question: Should we try to have the API properly reflect what it's actually configuring? i.e. `default` vs. default value? Maybe it's too late fix it properly and comments are sufficient.

@Miciah since we've talked a little about this before, do you have an opinion?

Reply:
@gcs278 that is why i have mentioned about this here https://github.com/openshift/enhancements/pull/1593/files#diff-cfc19a244868c1f04e3b6abf6779c80f4938e934d4fb8913a4b72ada48b266aeR125

Reply:
Right, I see that. My question is more of: is it possible the API structure/naming could somehow reflect what it's trying to configure? e.g. `spec.loadBalancer.platform.aws.nlbParameters.defaultIngressControllerEIPAllocations`

But on second though, that's pretty ugly, and I don't know if that makes sense. I don't have a good answer. I just thought it was worth pointing out the divergence of the API here. The API seems a bit inconsistent, but I suppose we don't have any better options other than just document it in the GO docs.